### PR TITLE
(feat) Hot reload Implementation (7/N) - Add ReplaceCluster operation to support modified cluster reload

### PR DESCRIPTION
**Pull request by Uzziee**
### Type of change

- Enhancement / new feature

### Description

As per the [discussed sequence](https://github.com/kroxylicious/kroxylicious/pull/3977#discussion_r3270313081), I have added a new `ReplaceCluster` operation which will handle reloading of a modified cluster.

### Additional Context

https://github.com/kroxylicious/kroxylicious/issues/4000

### Checklist

The checklist is used to help Committers judge a PR's readiness for merge.
Contributors: please check off items you consider done.

- [x] New tests written (where applicable).
- [ ] If making a user-facing change, documentation is provided, or if the intent is to provide documentation in a follow up PR, an issue is raised tracking the need for the documentation update. Link the to issue in this checklist. 
- [x] Existing unit/integration/system tests passing.
- [ ] Any Sonarcloud warnings are addressed (or suppressed with `@SuppressWarnings` and a justifying comment).
- [ ] PR references related GitHub issue(s) so they are closed on merging.
- [ ] If AI tools assisted with code changes, ensure commit messages include `Assisted-by:` trailer (see [DEV_GUIDE.md](../blob/main/DEV_GUIDE.md#ai-disclosure-requirement)).
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

### CODE REVIEW

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public VirtualClusterRegistry(List<VirtualClusterModel> virtualClusterModels,
                                   BiConsumer<String, Optional<Throwable>> onVirtualClusterStopped) {
         Objects.requireNonNull(virtualClusterModels, "virtualClusterModels must not be null");
         this.onVirtualClusterStopped = Objects.requireNonNull(onVirtualClusterStopped, "onVirtualClusterStopped must not be null");
-        this.entriesByCluster = new LinkedHashMap<>();
+        this.entriesByCluster = new ConcurrentHashMap<>();
```
Comment (Uzziee) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java]: I added this map in my last [PR](https://github.com/kroxylicious/kroxylicious/pull/4020), I gave some more thought on the choice of the map implementation.

I wanted to improve the threading model for this

### Threading Model

Writes happen exclusively on the reconfigure path, protected by `ConfigurationReloadOrchestrator`'s `reconfigureLock`.

Reads additionally occur on Netty event-loop threads via:
- `registerConnection()`
- `deregisterConnection()`

These methods are invoked when client connections are established or closed.

A `ConcurrentHashMap` is used to support:
- Lock-free reads
- Thread-safe writes

which matches the required access pattern.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java [hunk 2]
```diff
@@ -436,12 +437,165 @@ void shouldRemoveRuntimeAddedPortAddressedVc(@BrokerCluster KafkaCluster cluster
...
+                            .as("ReconfigureResult should have no errors for a port-relocation modify")
+                            .isFalse());
+
+            // The old port is reclaimable from outside the proxy — the remove half of the
+            // replace freed the binding.
+            assertPortIsBindable(oldPort);
+
+            // Drop the cached clients built around the old bootstrap so the next
+            // tester.producer/consumer call rebuilds against the new port.
+            tester.closeClientsFor("vc-relocate");
```
Comment (Uzziee) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java:528]: Added this new interface API, which lets us close the existing clients for a given VC.
During a reload for a modified cluster, if a port changes, the cached tester clients will continue to point to older closed port.
This API allows us to clear all the cached clients which allows us to register new clients, against the new VC port addresses.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/OperationsPlanner.java [hunk 3]
```diff
@@ -39,42 +43,65 @@ final class OperationsPlanner {
...
-                            "OperationsPlanner: no model for removed cluster '" + name
-                                    + "'; this indicates a ChangeDetector contract violation"
-                                    + " (cluster reported as removed but absent from the registry)");
-                }
-                ops.add(new RemoveCluster(model, virtualClusterRegistry, endpointRegistry));
+        // Resolve each model map at most once, and only when at least one phase needs it.
+        // A pure-remove plan doesn't pay for resolving newConfig (and vice versa). Modifies
+        // need both.
+        Map<String, VirtualClusterModel> registryModelsByName = (changes.clustersToRemove().isEmpty()
+                && changes.clustersToModify().isEmpty()) ? Map.of() : registryModelsByName();
```
Comment (SamBarker) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/OperationsPlanner.java]: Design note: `virtualClusterModels()` is being used here as a name-resolution mechanism — fetch all models, build a map, do `map.get(name)`. This is Tell-Don't-Ask: the planner is pulling data out of the registry and doing the lookup itself, work the registry could own.

A `modelFor(String clusterName)` method on `VirtualClusterRegistry` would let the planner state its intent directly. The planner always has a name from the `ChangeResult` and just needs the corresponding model — a direct lookup is the right shape. This would also free `virtualClusterModels()` to serve a single clear purpose: "what is currently configured and running," filterable to non-`STOPPED` entries without breaking the planner.

Not a blocker, but worth a follow-up before more callers grow different expectations about what `virtualClusterModels()` includes.

File: kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java [hunk 4]
```diff
@@ -327,6 +327,24 @@ public CompletableFuture<ReconfigureResult> reconfigure(Configuration newConfig)
         });
     }
 
+    @Override
+    public void closeClientsFor(String virtualCluster) {
```
Comment (SamBarker) [inline kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java:331]: There are unit tests for `DefaultKroxyliciousTester` — could you add coverage for this new method there? Useful cases would include: evicting only the named cluster's clients while leaving other clusters' clients untouched, and surfacing a close failure as an `IllegalStateException`.

**Review by SamBarker — APPROVED** [review]
Thanks — each PR in this series is getting easier to review as the abstractions settle. The operation model is reading cleanly and the test coverage for the edge cases (bind failure, port reuse, modify with port change) is solid.

One thing to be aware of: `shouldReleasePortWhenPortAddressedVcIsRemoved` is showing intermittent failures in CI with `EADDRINUSE` on port 51000/51010 at startup. This looks like a CI environment port conflict rather than a test logic issue — there's a separate PR planned to clean up the port allocation strategy across the hot-reload ITs, so not a blocker here.

See inline comments for two minor follow-ups.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/OperationsPlanner.java (hunk 3 above)
Comment (Uzziee) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/reload/OperationsPlanner.java]: Fair point.
I have added `modelFor()` method in VCR, using which we can get the existing (old) models.
However, we would still need to build the map for new models (which are to be added)

File: kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java (hunk 4 above)
Comment (Uzziee) [reply kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java:331]: Added unit tests for this new API

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java (hunk 2 above)
Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java:528]: Makes sense to me.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java
Comment (k-wall) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: (non-blocking) nit: we have the unwritten (and unenforced) convention that fields are at the top of the class.

File: kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java [hunk 5]
```diff
@@ -327,6 +327,24 @@ public CompletableFuture<ReconfigureResult> reconfigure(Configuration newConfig)
...
+                closeCloseable(e.getValue()).ifPresent(exceptions::add);
+                return true;
+            }
+            return false;
+        });
+        if (!exceptions.isEmpty()) {
+            // Log each close-failure (so a stuck client doesn't hide behind the rest) and
+            // surface the first as the test-visible failure. Matches the close() pattern.
+            exceptions.forEach(ex -> LOGGER.error(ex.getMessage(), ex));
+            throw new IllegalStateException(exceptions.get(0));
```
Comment (k-wall) [inline kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java]: nit: let's make it obvious to the programmer that >1 may have failed to close. something like (untested):

```suggestion
            throw new IllegalStateException("%d client(s) failed to close cleanly. First client close chained, others have been logged at ERROR".formatted(exceptions.size()), exceptions.get(0));
```

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java [hunk 6]
```diff
@@ -279,52 +294,38 @@ public CompletableFuture<Void> removeVirtualCluster(String clusterName) {
...
+     *         draining, or failed) cluster would be a contract violation.
      */
     public CompletableFuture<Void> addVirtualCluster(VirtualClusterModel newModel) {
         Objects.requireNonNull(newModel, "newModel must not be null");
         String name = newModel.getClusterName();
-        if (entriesByCluster.containsKey(name)) {
-            throw new IllegalArgumentException("Cluster already exists: " + name);
+        var existing = entriesByCluster.get(name);
+        if (existing != null && !(existing.lifecycle().state() instanceof VirtualClusterLifecycleState.Stopped)) {
+            throw new IllegalArgumentException("Cluster already exists and is not Stopped: " + name);
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java]: IAE seems questionable.  ISE might be a more natural choice.
should we report its actual state in the exception mechanism.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java [hunk 7]
```diff
@@ -83,7 +83,10 @@ public VirtualClusterRegistry(List<VirtualClusterModel> virtualClusterModels,
      * Returns the currently-tracked virtual cluster models. The list reflects the constructor-
      * supplied models PLUS any added at runtime via {@link #addVirtualCluster(VirtualClusterModel)}.
      *
-     * @return snapshot of currently-tracked virtual cluster models (insertion order)
+     * <p>Iteration order is unspecified (the backing map is concurrent). Callers that need
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java:86]: "Iteration order is unspecified (the backing map is concurrent)." -> so I think we should make the return type a Collection rather than a List, so this is implicit in the contract.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java (hunk 1 above)
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java]: nit: let's push the initialisation to the field.

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java
Comment (Uzziee) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Done

File: kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java (hunk 5 above)
Comment (Uzziee) [reply kroxylicious-integration-test-support/src/main/java/io/kroxylicious/testing/integration/tester/DefaultKroxyliciousTester.java]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java (hunk 6 above)
Comment (Uzziee) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java (hunk 7 above)
Comment (Uzziee) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java:86]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java (hunk 1 above)
Comment (Uzziee) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java]: Done

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java [hunk 8]
```diff
@@ -279,52 +294,38 @@ public CompletableFuture<Void> removeVirtualCluster(String clusterName) {
...
+     * @param newModel the model for the new cluster
      * @return an already-completed future (the operation is synchronous; the
      *         {@link CompletableFuture} shape is preserved for caller symmetry with
      *         {@link #removeVirtualCluster})
-     * @throws IllegalArgumentException if a cluster with this name already exists
+     * @throws IllegalArgumentException if an entry already exists AND its lifecycle is in any
+     *         state OTHER than {@code Stopped} — re-adding an actively-serving (or initializing,
+     *         draining, or failed) cluster would be a contract violation.
      */
     public CompletableFuture<Void> addVirtualCluster(VirtualClusterModel newModel) {
```
Comment (k-wall) [inline kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java:319]: Claude suggested to me that we ought to avoid the check/act pattern.  The VCR doesn't expect to be mutated by multiple thread, but I still prefer its suggestion.

```
  Option 1: Use compute() for Atomic Check-and-Update (Recommended)

  public CompletableFuture<Void> addVirtualCluster(VirtualClusterModel newModel) {
      Objects.requireNonNull(newModel, "newModel must not be null");
      String name = newModel.getClusterName();

      entriesByCluster.compute(name, (key, existing) -> {
          // This lambda executes atomically for this key
          if (existing != null && !(existing.lifecycle().state() instanceof VirtualClusterLifecycleState.Stopped))
  {
              throw new IllegalArgumentException("Cluster already exists and is not Stopped: " + name);
          }

          LOGGER.atInfo()
                  .addKeyValue("virtualCluster", name)
                  .addKeyValue("operation", "addVirtualCluster")
                  .addKeyValue("replacingStoppedEntry", existing != null)
                  .log("reconfigure: created lifecycle in INITIALIZING; gateway registration is the orchestrator's
  responsibility");

          return new VirtualClusterEntry(newModel, new VirtualClusterLifecycle(name, newModel.drainTimeout()));
      });

      return CompletableFuture.completedFuture(null);
  }
```

**Review by k-wall — APPROVED** [review]
Couple of nits and suggestions but otherwise LGTM.

File: kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java (hunk 8 above)
Comment (Uzziee) [reply kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/VirtualClusterRegistry.java:319]: Have made this change

File: kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java [hunk 9]
```diff
@@ -75,6 +76,32 @@ class HotReloadIT extends BaseIT {
...
+    private static final int PORT_BLOCK_BASE = 20000 + ThreadLocalRandom.current().nextInt(8000);
+    private static final int PORT_STRIDE = 10;
+    private static final int PORT_BLOCK_REMOVE = PORT_BLOCK_BASE; // shouldReleasePortWhenPortAddressedVcIsRemoved
+    private static final int PORT_BLOCK_ADD = PORT_BLOCK_BASE + 100; // shouldStartServingAddedPortAddressedVcEndToEnd
+    private static final int PORT_BLOCK_BINDFAIL = PORT_BLOCK_BASE + 200; // shouldSurfaceBindFailureAsReconfigureError...
+    private static final int PORT_BLOCK_REUSE = PORT_BLOCK_BASE + 300; // shouldSupportPortReuseAcrossReconfigures
+    private static final int PORT_BLOCK_ADD_THEN_REMOVE = PORT_BLOCK_BASE + 400; // shouldRemoveRuntimeAddedPortAddressedVc
+    private static final int PORT_BLOCK_MODIFY_SAME_PORT = PORT_BLOCK_BASE + 500; // shouldModifyPortAddressedVcWithSamePort
+    private static final int PORT_BLOCK_MODIFY_DIFF_PORT = PORT_BLOCK_BASE + 600; // shouldModifyPortAddressedVcWithDifferentPort
+    private static final int PORT_BLOCK_MODIFY_FAIL = PORT_BLOCK_BASE + 700; // shouldSurfaceModifyFailureAsReconfigureError
```
Comment (Uzziee) [inline kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Hi @k-wall , I wanted to try this out in this PR.
Maybe if we randomize the port allocation and chose ports outside of ephemeral range, the flaky issue might get resolved
This is related to - https://github.com/kroxylicious/kroxylicious/issues/4065#issuecomment-4620190269

Lets see how the test run goes

Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Please see my proposed fix on #4067. I don't believe you'll need this.

Comment (Uzziee) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Ah ok, lets wait for your build to pass the IT test.
If it works, I can revert this change (Though I think the current state is also good, randomizing ports) wdyt ?
I might have to update the comments though

Comment (Uzziee) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: @k-wall The tests ran successfully in your PR.

Do you want me to revert back to hardcoding of ports or keep this randomization in place ?

Comment (k-wall) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Let's revert the randomization. Let's land https://github.com/kroxylicious/kroxylicious/pull/4067 (@SamBarker would you be able to review and merge) then we can rebase this PR on top and land this one too.

Comment (Uzziee) [reply kroxylicious-integration-tests/src/test/java/io/kroxylicious/it/HotReloadIT.java]: Reverted

**Comment by Uzziee** [comment]
@k-wall @SamBarker After merging the port REUSE PR - https://github.com/kroxylicious/kroxylicious/pull/4067

I am still seeing the same failures in HotReloadIT .... 
I am thinking to re-add that randomization and chose ports below the ephemeral range, any concerns ?

**Comment by k-wall** [comment]
> @k-wall @SamBarker After merging the port REUSE PR - #4067
> 
> I am still seeing the same failures in HotReloadIT .... I am thinking to re-add that randomization and chose ports below the ephemeral range, any concerns ?

Ok, let do that as a tactical fix to let us move forward.

**Comment by Uzziee** [comment]
@k-wall @SamBarker the tests passed after adding randomization code, can you help merge this now ?

**Review by k-wall — APPROVED** [review]
LGTM
